dhall-docs/src/Dhall/Docs/Core.hs

```diff
@@ -0,0 +1,232 @@
+    2. @\/a\/c\/index.html@ that will list the @\/a\/c\/b.txt@ file
+
+-}
+createIndexes
+    :: Path Abs Dir    -- ^ Directory where index.html file will be saved. Used
+                       --   to link the css resources. It should be a prefix for
+                       --   each @Path Abs File@ on the second argument
+    -> [Path Abs File] -- ^ Html files generated by the tool
+    -> String          -- ^ Package name
+    -> IO ()
+createIndexes outputPath htmlFiles packageName = do
+    let toMap file = Map.singleton (Path.parent file) [file]
+    let filesGroupedByDir = Map.unionsWith (<>) $ map toMap htmlFiles
+
+    let listDirRel dir = do
+            dirs <- fst <$> Path.IO.listDir dir
+            mapM (Path.stripProperPrefix dir) dirs
+
+    let createIndex index files = do
+            indexFile <- Path.fromAbsFile . (index </>) <$> Path.parseRelFile "index.html"
+            indexTitle <-
+                if outputPath == index then return "package"
+                else Path.fromRelDir <$> Path.stripProperPrefix outputPath index
+            indexList <- Control.Monad.forM files $
+                fmap Path.filename . Path.stripProperPrefix outputPath
+            dirList <- listDirRel index
+
+            let relativeResourcesPath = resolveRelativePath outputPath index
+            Lucid.renderToFile indexFile $
+                indexToHtml
+                    indexTitle
+                    indexList
+                    dirList
+                    DocParams
+                        { relativeResourcesPath = relativeResourcesPath
+                        , packageName = packageName}
+
+    _ <- Map.traverseWithKey createIndex filesGroupedByDir
+    return ()
+
+-- | Generate all of the docs for a package
+generateDocs
+    :: Path Abs Dir  -- ^ Input directory
+    -> Path Abs File -- ^ Link to be created to the generated documentation
+    -> String        -- ^ Package name, used in some HTML titles
+    -> IO ()
+generateDocs inputDir outLink packageName = do
+    outDir <- Applicative.liftA2 (</>) getDocsHomeDirectory (Path.parseRelDir packageName)
+
+    dhallFilesAndHeaders <- getAllDhallFilesAndHeaders inputDir
+    if null dhallFilesAndHeaders then
+        putStrLn $
+            "No documentation was generated because no file with .dhall " <>
+            "extension was found"
+    else do
+        generatedHtmlFiles <-
+            mapM (uncurry $ saveHtml inputDir outDir packageName)
+                dhallFilesAndHeaders
+        createIndexes outDir generatedHtmlFiles packageName
+
+        dataDir <- getDataDir
+        Path.IO.ensureDir outDir
+        Control.Monad.forM_ dataDir $ \(filename, contents) -> do
+            let finalPath = Path.fromAbsFile $ outDir </> filename
+            Data.ByteString.writeFile finalPath contents
+
+        Path.IO.ensureDir $ Path.parent outLink
+        Path.IO.createFileLink
+            (outDir </> Data.Maybe.fromJust (Path.parseRelFile "index.html"))
```

Comment:
I generally recommend against using partial functions like `fromJust`.  I like to get into the habit of handling all errors with good error messages, even for errors that I believe are impossible (if only to incentivize myself to make invalid states unrepresentable)

Reply:
I thought it was ok since it is impossible for `Path.parseRelFile "index.html"` to be `Nothing.

The best we can do here is use the `path` quasiquoters, but on hydra it complains about a missing import (`Rel` from `path) and on appveyor it requires it. I see that appveyor uses `stack`, which uses `ghc-8.6.5`, unlike hydra, that uses `8.6.1`. May this be a bug on ghc or I misunderstood something?

Reply:
@german1608 Can you give us links to the failing jobs?

It's a weird issue, and this is not a very pretty workaround. (I see you're using `parse{Abs,Rel}{File,Dir}` on literals elsewhere too).

Did you try using template haskell instead of the quasiquoters?

Reply:
A better alternative could be to disable the warning for the problematic GHC version with a bit of CPP.

Reply:
I'll replicate the issues on this PR and sent here the links to builds, gimme a sec

Reply:
I didn't tried using template haskell though... I'll give it a try after replicating the errors

Reply:
Commit https://github.com/dhall-lang/dhall-haskell/pull/1876/commits/49e59f3b21e1422e03b6942c9ed8fc9f82c639e9 uses quasiquoters, and imports `Rel`.

Its hydra build [fails](https://hydra.dhall-lang.org/build/62671/nixlog/2) because of the redundant import

Its appveyor build [succeeds](https://ci.appveyor.com/project/Gabriel439/dhall-haskell/builds/33699055)

Commit https://github.com/dhall-lang/dhall-haskell/pull/1876/commits/ebb8a3664c2db4e5e871dbd82108ae513a9c70e1 uses quasiquoters, but doesn't import `Rel`.

Its hydra build [succeeds](https://hydra.dhall-lang.org/build/62677)

Its appveyor build [fails](https://ci.appveyor.com/project/Gabriel439/dhall-haskell/builds/33699326) because it can't found `Rel` on scope.

Reply:
It also fails using template haskell :(

![image](https://user-images.githubusercontent.com/29154504/85470618-4a481980-b57d-11ea-82d1-183466c86752.png)

Reply:
Thanks, @german1608, strange indeed! I wasn't aware that quasiquoters require additional imports! If you're using a *nix OS, can you reproduce the build failure with 8.6.5?

If the `Rel` import is in fact needed, I'd suggest to add `{-# OPTIONS_GHC -Wno-redundant #-}` to the problematic module (possibly behind CPP for the specific GHC version), or not to use `-Werror` by removing this line:

https://github.com/dhall-lang/dhall-haskell/blob/6e1fdd8696856660d74b9eeb41e8a91249d0c8cf/nix/shared.nix#L110

Reply:
…or use `-Wwarn` in the specicific module to negate `-Werror`… http://downloads.haskell.org/~ghc/latest/docs/html/users_guide/using-warnings.html#ghc-flag--Wwarn

Reply:
With *nix you mean NixOS? I'm using ubuntu and `stack` is using `lts-14.27`, which is 8.6.5 already.

Yeah, a solution is to use some CPP but its strange for me that is failing on a greater version.

Reply:
No, I was wondering whether you could reproduce the issue with a non-Windows OS, i.e. macOS or Linux, some Unix derivative.



> its strange for me that is failing on a greater version.

What do you mean?

Reply:
> No, I was wondering whether you could reproduce the issue with a non-Windows OS, i.e. macOS or Linux, some Unix derivative.

Oh I see, well, as I said, I tested on ubuntu. The `-Wwarn` looks like it will work, but I don't like the idea of suppressing warnings. So I'll go with the CPP approach although I don't like to (ab)use it.

> > its strange for me that is failing on a greater version.
> 
> What do you mean?

I mean that nix uses ghc 8.6.1 and it doesn't fail without the import, while stack (appveyor) uses 8.6.5 and it does, which is a greater version AFAIK. That it why its weird for me, although there may a bug in that newer version... or maybe I'm confusing stuff, first time I used haskell pragmas was on this GSoC haha.